dataclasses_json/core.py

```diff
@@ -314,14 +314,16 @@ def _decode_generic(type_, value, infer_missing):
             type_options = _get_type_args(type_)
             res = value  # assume already decoded
             if type(value) is dict and dict not in type_options:
-                # FIXME if all types in the union are dataclasses this
-                #  will just pick the first option -
-                #  maybe find the best fitting class in that case instead?
+                changed = False
                 for type_option in type_options:
                     if is_dataclass(type_option):
-                        res = _decode_dataclass(type_option, value, infer_missing)
-                        break
-                if res == value:
+                        try:
+                            res = _decode_dataclass(type_option, value, infer_missing)
+                            changed = True
+                            break
+                        except (KeyError, ValueError):
```

Comment:
It is unclear to me what is the real life case for this, and why do we skip here instead of failing?

Reply:
This except block catches two situations:

1.  When the dataclass in consideration does not contain one (or more) fields that exist in the `value`.
2.  When the type of one (or more) of the fields that exists in the `value` are not unmarshallable into the type of the corresponding field.

If either of these exceptions are thrown, then we know that the dataclass under consideration is not an appropriate choice to decode into - and we continue onto the next type in the `type_options` array.

This code is exercised in the test `test_deserialize_without_discriminator` - where json strings that do not contain `__type` fields are successfully decoded into the correct types.

The real life use case for this is trying to deserialize data that do not have the ___type discriminator value specified.  This could occur when receiving data that was not encoded by the dataclasses_json project, from an HTTP endpoint for example.

Reply:
Understood, thanks for explanation!